src/typeInference/annotateAST.ts

```diff
@@ -77,6 +60,9 @@ export function annotate(ast: IAST): SequenceType | undefined {
 			]);
 			return { type: ValueType.XSSTRING, mult: SequenceMultiplicity.EXACTLY_ONE };
 		default:
+			for (let i = 1; i < ast.length; i++) {
+				annotate(ast[i] as IAST);
```

Comment:
Can we make sure that it won't be stuck in a recursion loop?

Reply:
The whole test suit passes and the parser will never generate circular AST trees so this shouldn't be a problem.

Reply:
Okay cool well in that case it should be fine